internal/build/build.go

```diff
@@ -0,0 +1,266 @@
+		goodNewsEmoji        = "✨"
+		checksPassedEmoji    = "✅"
+		mbStringFormatter    = "%.1fMB"
+	)
+
+	// build example binary
+	err = runCmd("go", "build", "-o", builtFilePath, sourceFilePath)
+	if err != nil {
+		return err
+	}
+
+	// get file into
+	fileInfo, err := os.Stat(builtFilePath)
+	if err != nil {
+		return err
+	}
+
+	// get human readable size, in MB with one decimal place.
+	// example output is: 35.2MB. (note: this simply an example)
+	// that output is much easier to reason about than the `35223432`
+	// that you would see output without the rounding
+	fileSize := fileInfo.Size()
+	fileSizeInMB := float64(fileSize) / float64(1000000)
+	roundedFileSize := math.Round(fileSizeInMB*10) / 10
+	roundedFileSizeString := fmt.Sprintf(mbStringFormatter, roundedFileSize)
+
+	// check against bounds
+	isLessThanDesiredMin := roundedFileSize < desiredMinBinarySize
+	isMoreThanDesiredMax := roundedFileSize > desiredMaxBinarySize
+	desiredMinSizeString := fmt.Sprintf(mbStringFormatter, desiredMinBinarySize)
+	desiredMaxSizeString := fmt.Sprintf(mbStringFormatter, desiredMaxBinarySize)
+
+	// show guidance
+	fmt.Println(fmt.Sprintf("\n%s is the current binary size", roundedFileSizeString))
+	// show guidance for min size
+	if isLessThanDesiredMin {
+		fmt.Println(fmt.Sprintf("  %s %s is the target min size", goodNewsEmoji, desiredMinSizeString))
+		fmt.Println("") // visual spacing
+		fmt.Println("     The binary is smaller than the target min size, which is great news!")
+		fmt.Println("     That means that your changes are shrinking the binary size.")
+		fmt.Println("     You'll want to go into ./internal/build/build.go and decrease")
+		fmt.Println("     the desiredMinBinarySize, and also probably decrease the ")
+		fmt.Println("     desiredMaxBinarySize by the same amount. That will ensure that")
+		fmt.Println("     future PRs will enforce the newly shrunk binary sizes.")
+		fmt.Println("") // visual spacing
+		os.Exit(1)
+	} else {
+		fmt.Println(fmt.Sprintf("  %s %s is the target min size", checksPassedEmoji, desiredMinSizeString))
+	}
+	// show guidance for max size
+	if isMoreThanDesiredMax {
+		fmt.Println(fmt.Sprintf("  %s %s is the target max size", badNewsEmoji, desiredMaxSizeString))
+		fmt.Println("") // visual spacing
+		fmt.Println("     The binary is larger than the target max size.")
+		fmt.Println("     That means that your changes are increasing the binary size.")
+		fmt.Println("     The first thing you'll want to do is ask your yourself")
+		fmt.Println("     Is this change worth increasing the binary size?")
+		fmt.Println("     Larger binary sizes for this package can dissuade its use.")
+		fmt.Println("     If this change is worth the increase, then we can up the")
+		fmt.Println("     desired max binary size. To do that you'll want to go into")
+		fmt.Println("     ./internal/build/build.go and increase the desiredMaxBinarySize,")
+		fmt.Println("     and increase the desiredMinBinarySize by the same amount.")
```

Comment:
We probably want to add more details about what the binary size increases, like:
- Added dependencies
- Added const values (.text, .rodata)
- What else?

Reply:
I think if anything this console message should get smaller (because it's annoying to update) and we should point to a long-ish thing in the `docs/` folder that describes all the various ways binary size could increase. I'm imagining there being a `docs/developement.md#binary-size` section that's like 10 paragraphs long 😄 

Does that make sense? Or should we put something shorter into this error message?

@saschagrunert

Reply:
SGTM :)